src/compiler/checker.ts

```diff
@@ -35601,10 +35603,11 @@ namespace ts {
                 // Grammar checking
                 if (catchClause.variableDeclaration) {
                     const declaration = catchClause.variableDeclaration;
-                    if (declaration.type) {
-                        const type = getTypeForVariableLikeDeclaration(declaration, /*includeOptionality*/ false);
+                    const typeNode = getEffectiveTypeAnnotationNode(getRootDeclaration(declaration));
+                    if (typeNode) {
+                        const type = getTypeFromTypeNode(typeNode);

```

Comment:
I think you should call getTypeForVariableLikeDeclaration here instead of getTypeFromTypeNode. If that doesn't work right now, then getTypeOfVariableOrParameterOrPropertyWorker should be fixed.